File: python env and code/code1.py
import numpy as np
import random
#import multiprocessing
from joblib import Parallel, delayed
import matplotlib.pyplot as plt


def hardselect(n,p):
    size = int(np.floor(n*p))
    v = np.array(range(0,n))
    taboo = v[:size]
    return(taboo)

def softselect(M,n,p):
    size = int(np.floor(n*p))
    v = np.array(range(0,n))
    grand_prob = np.sum(M, axis = 1)
    prob_vec = grand_prob[0:n]/np.sum(grand_prob[0:n])
    taboo = np.random.choice(v, size = size, replace = False, p = prob_vec)
    return(taboo)

def simulate_graph_maximal(N,p):
    M = np.zeros(shape = (N,N), dtype = int)
    M[0,1] = 1; M[1,0] = 1
    for i in range(2,N):
        taboo = hardselect(i, p)
        v = range(0, i)
        nontaboo = np.setxor1d(v, taboo, assume_unique=True)
        nontaboo_incidence = M[nontaboo]
        prob_vec = np.sum(nontaboo_incidence, axis = 1)
        edge_end = random.choices(nontaboo, weights = prob_vec, k = 1)
        M[i, edge_end] = 1
        M[edge_end, i] = 1
    
    #deg_seq = np.sum(M, axis = 1)
    max_deg = max(np.sum(M, axis = 1))

    return max_deg

def simulate_graph_seq(N,p,k):
    M = np.zeros(shape = (N,N), dtype = int)
    M[0,1] = 1; M[1,0] = 1
    for i in range(2,N):
        taboo = hardselect(i, p)
        v = range(0, i)
        nontaboo = np.setxor1d(v, taboo, assume_unique=True)
        nontaboo_incidence = M[nontaboo]
        prob_vec = np.sum(nontaboo_incidence, axis = 1)
        edge_end = random.choices(nontaboo, weights = prob_vec, k = 1)
        M[i, edge_end] = 1
        M[edge_end, i] = 1
    
    deg_seq = np.sum(M, axis = 1)
    number_of_vertex = deg_seq.count(k)

    return number_of_vertex

def maximal_plot(N,p,S):
    #num_cores = multiprocessing.cpu_count()
    inputs = range(0,S)

    mylist = []
    mylist = Parallel(n_jobs = -1)(delayed(simulate_graph_maximal)(N,p) for i in inputs)
    fig, ax = plt.subplots(figsize =(10, 7))
    ax.hist(mylist)
    plt.show()

maximal_plot(500, .1, 5)

File: R code/code1.R
N = 1000; p = 0.1
M = matrix(0, nrow = N, ncol = N)
M[1,2] = 1; M[2,1] = 1


hardselect = function(n,p) {
  size = floor(n*p)
  v = c(1:n)
  if(size == 0) {
    return(integer(0))
  }
  else {
    taboo = v[1:size]
    return(taboo)
  }
}

softselect = function(M,n,p) {
  size = floor(n*p)
  v = c(1:n)
  grand_prob = rowSums(M)
  prob_vec = grand_prob[1:n]
  taboo = sample(v, size = size, replace = F, prob = prob_vec)
  return(taboo)
}

for(i in 3:N) {
  #set.seed(i)
  taboo = hardselect((i-1), p)
  v = c(1:(i-1))
  nontaboo = v[!v %in% taboo]
  nontaboo_incidence = M[nontaboo, ]
  prob_vec = rowSums(M[nontaboo, ])
  edge_end = sample(nontaboo, 1, prob = prob_vec)
  M[i, edge_end] = 1
  M[edge_end, i] = 1
}
#hist(rowSums(M), freq = F)
print(max(rowSums(M)))